tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py

```diff
@@ -175,53 +177,101 @@ def fit(self, x: pd.DataFrame, y=None, **fit_params) -> "Method":
         self._hashtags_tf_idf_vectorizer = TfidfVectorizer(norm="l2")
         self._hashtags_tf_idf_representation = self._hashtags_tf_idf_vectorizer.fit_transform(document_list)
 
+        if self.verbose:
+            print("Fitting hashtag encoders ...")
+
+        # [:-4] because each hashtag has "_tag" postfix to distinguish it in the graph
+        self._hashtag_encoder.fit([lab[:-4] for lab in self._hashtag_labels])
+
         return self
 
-    def transform(self, x: Union[List[str], str]) -> List[str]:
+    def transform(self, x: Union[Collection[Collection[str]], Collection[str]], **kwargs) -> np.ndarray:
```

Comment:
Nie jestem pewien czy `Collection` pozwala na odwołanie do elementu za pomocą operatora `[ ]`. Nie jestem też pewien czy to dla nas istotne :P

A tak na marginesie: czy byłby też ok zapis `Collection[Union[Collection[str], str]]`?

Reply:
Generalnie w typingu jest problem z tym niesamowity, żeby uwzględnić dowolny typ Indexable. I racja, `Collection` nie pozwala na to, ale chyba `Tuple` i `List` się pod tym mieszczą. Ludzie proponują zazwyczaj stworzenie własnego typu po prostu.

Taka jest średnia o tyle, że `Union` wskazuje na możliwość wystąpienia jednego z `N` typów pod nawiasami. W tym przypadku:
- `Collection[Collection[str]]` - lista `M` tweetów, gdzie każdy ma listę lematów, a każdy lemat to string
- `Collection[str]` - lista `M` niezlematyzowanych tweetów, czyli treść tweeta to pojedynczy string.